### ISSUE TITLE
Implement `eq` for `str` type

### ISSUE THREAD

**Issue by emilyaherbert** [FuelLabs/sway#1060]

**Comment by nfurfaro** [comment]
I was looking to pick this up as I'm in need of a project. Not sure how to approach the fact (as mentioned above) that every different length of `str` is its own type.
A simple generic function could be a simple approach, but I would prefer to implement the existing `Eq` trait if at all possible. It seems like we would need maybe trait constraints to make this work?
ie: `impl Eq for str[n]` where `n` can be any u64 .

**Comment by otrho** [comment]
Internally strings of different length are definitely of different types.  So using `Eq` would enforce that both strings are the same length at the type checker.  But yeah, we need higher level 'kinds' to match strings of all lengths/types for a single `impl`.  Unfortunately this feels like we might need some compiler special magic to special case them in the type checker.

As per my comment above to actually compare them you'd use `MEQ` (or the fabled intrinsics which I swear will happen one day) but you need a length.

It would make sense for `size_of` to return the byte size of the strings and you could use that.  OTOH there's still the UTF8 issue which needs resolving.  Although both strings may have the same number of characters they could have very different byte sizes.  So perhaps getting the `size_of` both strings, comparing sizes and then contents is the go.

**Comment by otrho** [comment]
The problem above is that `T` could be anything.  We really need to change something in the type system to make this work, AFAICT.  Either we make `str` a type of string of any fixed length, or we introduce a trait for strings and only strings which gives us `.num_chars()` and `.num_bytes()` and can also be used in `where` clauses.

```rust
pub fn are_strings_eq<S>(a: S, b: S) -> bool
    where S: StaticString
{
    ...
}
```
Do we have `where` clauses for functions yet, or just type declarations?  That example is really clumsy either way.

@sezna This is where your expertise shines.  How shall we use strings of unknown lengths with type safety?

**Comment by sezna** [comment]
We don't provide any native support for dynamically sized strings. That's Library Territory™.

 We'd either need to introduce a mechanism for implementing traits over all strs, or, defer to libraries. I vote for the latter because higher level string support is not an explicit goal of Sway and is not generally a high priority for smart contract development.

**Comment by otrho** [comment]
Note that we're not talking about dynamically sized strings at all here.  These problems are with Sway's static `str`.

So you're conceding that `str` as we have them now are essentially useless and we should instead construct them using libraries.  We might as well remove `str` as a basic type altogether then.

**Comment by otrho** [comment]
Except it really makes sense to keep string literals.  I honestly think we need to tone down the type system and copy Rust to a degree.

We should make `str` a static const string of any fixed length.  The length should not be a part of the type.  The have a byte length property which we can fetch with `size_of` (and optionally wrap in `.len()` in a library).  They can't be indexed.  UTF8 chars can be managed in libraries once we have iterators.

Thoughts?

**Comment by adlerjohn** [comment]
>  They have a byte length property which we can fetch with `size_of`

Is that not the case currently?

**Comment by otrho** [comment]
Yep, probably?  I was just trying to lay out everything we need to make it all work.  @sezna Does the above seem fair?

**Comment by nfurfaro** [comment]
>Is that not the case currently?

I think so... I have WIP for this with tests, but blocked by `MemoryOverflow`.

**Comment by sezna** [comment]
It makes sense, but you're suggesting essentially keeping the restriction that all `str`s must have a size known at compile time, but they fall under the same type technically? That's pretty much exactly what Rust does, seems reasonable.

**Comment by otrho** [comment]
OK, I'm renaming this issue since all of the discussion has been here already.  This is a largish change though, as it should go right through from parsing to IR.

**Comment by sezna** [comment]
Feel free to override me, but I'm putting p: medium on this since strings are not blocking any major use cases right now

**Comment by xunilrj** [comment]
We have a new PR https://github.com/FuelLabs/sway/pull/4996 that changed the semantics of string literals and a new issue https://github.com/FuelLabs/sway/issues/5110 to address improvements to string slices.
So I will close this issue.
